README.md
# codice_fiscale

Package per generare il codice fiscale italiano a partire dai dati anagrafici.

Contiene il database dei comuni italiano con provincia e codice comune del catasto (city_dao.dart)

# Costruttore

```
final fiscalCode = CodiceFiscale(
    firstName: 'marco eliseo',
    lastName: 'sanson',
    birthCity: CityDao().getCityByName('Vicenza'),
    gender: CodiceFiscaleGender.M,
    birthDate: DateTime(1981, 10, 28));

print(fiscalCode.lastNameCode());
```

# Validazione

```
CodiceFiscale.check("ABCXYZ00E05C62BP")
```


CHANGELOG.md
## [0.0.2] - 2020.12.11

* Merged pull request from @mtiziano

## [0.0.1] - 2020.12.11

* This is the first release


test/codice_fiscale_test.dart
import 'package:flutter_codice_fiscale/dao/city_dao.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_codice_fiscale/codice_fiscale.dart';

void main() {
  test('test last name', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.lastNameCode(), 'SNS');
  });

  test('test first name', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.firstNameCode(), 'MCL');
  });

  test('test birth date Male', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.birthDateCode(), '81R28');
  });

  test('test birth date Female', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.F,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.birthDateCode(), '81R68');
  });

  test('test birth place vicenza', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.F,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.birthPlaceCode(), 'L840');
  });

  test('test birth place padova', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Padova')!,
        gender: CodiceFiscaleGender.F,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.birthPlaceCode(), 'G224');
  });

  test('test check code char one', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.getCheckCode(), 'T');
  });

  test('test check code char two', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Padova')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.getCheckCode(), 'G');
  });

  test('calculate male', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Vicenza')!,
        gender: CodiceFiscaleGender.M,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.fiscalCode, 'SNSMCL81R28L840T');
  });

  test('calculate female', () {
    final fiscalCode = CodiceFiscale(
        firstName: 'marco eliseo',
        lastName: 'sanson',
        birthCity: CityDao().getCityByName('Padova')!,
        gender: CodiceFiscaleGender.F,
        birthDate: DateTime(1981, 10, 28));
    expect(fiscalCode.fiscalCode, 'SNSMCL81R68G224K');
  });
}


lib/utils/constants.dart
const MONTH_CODES = ['A', 'B', 'C', 'D', 'E', 'H', 'L', 'M', 'P', 'R', 'S', 'T'];

const CHECK_CODE_ODD = {
  '0': 1,
  '1': 0,
  '2': 5,
  '3': 7,
  '4': 9,
  '5': 13,
  '6': 15,
  '7': 17,
  '8': 19,
  '9': 21,
  'A': 1,
  'B': 0,
  'C': 5,
  'D': 7,
  'E': 9,
  'F': 13,
  'G': 15,
  'H': 17,
  'I': 19,
  'J': 21,
  'K': 2,
  'L': 4,
  'M': 18,
  'N': 20,
  'O': 11,
  'P': 3,
  'Q': 6,
  'R': 8,
  'S': 12,
  'T': 14,
  'U': 16,
  'V': 10,
  'W': 22,
  'X': 25,
  'Y': 24,
  'Z': 23
};

const CHECK_CODE_EVEN = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4,
  '5': 5,
  '6': 6,
  '7': 7,
  '8': 8,
  '9': 9,
  'A': 0,
  'B': 1,
  'C': 2,
  'D': 3,
  'E': 4,
  'F': 5,
  'G': 6,
  'H': 7,
  'I': 8,
  'J': 9,
  'K': 10,
  'L': 11,
  'M': 12,
  'N': 13,
  'O': 14,
  'P': 15,
  'Q': 16,
  'R': 17,
  'S': 18,
  'T': 19,
  'U': 20,
  'V': 21,
  'W': 22,
  'X': 23,
  'Y': 24,
  'Z': 25
};

const CHECK_CODE_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';


lib/codice_fiscale.dart
library codice_fiscale;

import 'package:flutter_codice_fiscale/utils/constants.dart';
import 'package:flutter_codice_fiscale/utils/helpers.dart';

import 'models/city.dart';
import 'dao/city_dao.dart';

enum CodiceFiscaleGender { M, F }

class CodiceFiscale {
  String firstName, lastName;
  City? birthCity;
  CodiceFiscaleGender gender;
  DateTime birthDate;
  String? birthCityString;
  late String _fiscalCode;

  CodiceFiscale(
      {required this.firstName,
      required this.lastName,
      required this.birthCity,
      required this.gender,
      required this.birthDate})
      : assert(firstName.isNotEmpty),
        assert(lastName.isNotEmpty) {
    _fiscalCode = '${lastNameCode()}${firstNameCode()}${birthDateCode()}${birthPlaceCode()}${getCheckCode()}';
  }

  CodiceFiscale.searchcity(
      {required this.firstName,
      required this.lastName,
      required this.birthCityString,
      required this.gender,
      required this.birthDate})
      : assert(firstName.isNotEmpty),
        assert(lastName.isNotEmpty),
        assert(birthCityString != null && birthCityString.isNotEmpty) {
    this.birthCity = CityDao().getCityByName(birthCityString);
    _fiscalCode = '${lastNameCode()}${firstNameCode()}${birthDateCode()}${birthPlaceCode()}${getCheckCode()}';
  }

  static bool check(String codiceFiscale) {
    String cf = codiceFiscale.toUpperCase();
    if (cf.length != 16) {
      return false;
    }
    if (!RegExp(r'^[A-Z]{6}[0-9LMNPQRSTUV]{2}[ABCDEHLMPRST]{1}[0-9LMNPQRSTUV]{2}[A-Z]{1}[0-9LMNPQRSTUV]{3}[A-Z]{1}$')
        .hasMatch(cf)) {
      return false;
    }
    final expectedCheckCode = codiceFiscale[15];
    cf = codiceFiscale.substring(0, 15);
    return _getCheckCode(cf).toUpperCase() == expectedCheckCode.toUpperCase();
  }

  String get fiscalCode => _fiscalCode;

  // Useful methods

  String lastNameCode() {
    final String code = '${extractConsonants(this.lastName)}${extractVowels(this.lastName)}XXX';
    return code.substring(0, 3).toUpperCase();
  }

  String firstNameCode() {
    String code = extractConsonants(this.firstName);
    if (code.length >= 4) {
      code = '${code[0]}${code[2]}${code[3]}';
    } else {
      code += '${extractVowels(this.firstName)}XXX';
      code = code.substring(0, 3);
    }
    return code.toUpperCase();
  }

  String birthDateCode() {
    String year = '${this.birthDate.year.toString()[2]}${this.birthDate.year.toString()[3]}';
    String month = MONTH_CODES[this.birthDate.month - 1];
    int dayInt = this.birthDate.day;
    if (gender == CodiceFiscaleGender.F) {
      dayInt += 40;
    }
    String day = '0$dayInt';
    day = day.substring(day.length - 2, day.length);
    return '$year$month$day';
  }

  String birthPlaceCode() {
    return birthCity!.code;
  }

  static String _getCheckCode(String partialFiscalCode) {
    partialFiscalCode = partialFiscalCode.toUpperCase();
    int val = 0;
    for (int i = 0; i < 15; i = i + 1) {
      final String c = partialFiscalCode[i];
      val += (i % 2 != 0 ? CHECK_CODE_EVEN[c.toString()]! : CHECK_CODE_ODD[c.toString()]!);
    }
    val = val % 26;
    return CHECK_CODE_CHARS[val];
  }

  String getCheckCode() {
    String partialFiscalCode = '${lastNameCode()}${firstNameCode()}${birthDateCode()}${birthPlaceCode()}';
    return _getCheckCode(partialFiscalCode);
  }
}
